### PR TITLE
feat: add metaMask SDK connector

### PR DESCRIPTION
**Pull request by EdouardBougon**
# MetaMask SDK integration

This PR adds the connection to MetaMask using the MetaMask SDK.

The SDK must be used for both connecting via the extension and on mobile.

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @EdouardBougon , thank you for your contribution!

I've reviewed the PR and have some issues to report: 
1. Different networks when connect to MM: 
- open cowswap
- connect to MM on Sepolia (for example) --> the app sends a request to change a network to Ethereum. 
Expected: it should not force a user to connect to Ethereum. Sepolia details should be displayed on CoW Swap
change back
2. When another wallet extensions are enabled in a browser, they are duplicating in a connection modal
duplicating
3. Mobile: in 3 cases out of 4, when I connected to MM, I saw 'could not load balances' error message --> balances were not loaded
image
4. When I finally saw balances, I tried placing a trade, but a request was not sent to the wallet. Tested on iOS 17.6.1, Chrome+MM (WC)
5. (reproduced only once): I sent a request to change a network in MM ( iOS 17.6.1, Safari+MM (WC)), accepted in MM app, and the app started to redirect me to MM again and again until I closed it. See the video: 

https://github.com/user-attachments/assets/59fa629e-ce0a-41e7-a992-e153ac31db46

Could you please take a look at these issues?

File: apps/cowswap-frontend/src/modules/wallet/containers/ConnectWalletOptions.tsx [hunk 1]
```diff
@@ -57,19 +59,13 @@ interface InjectedOptionsProps {
...
-    if (multiInjectedProviders.length) {
-      return (
-        <>
-          {multiInjectedProviders.map((providerInfo) => {
+  if (multiInjectedProviders.length) {
+    return (
+      <>
+        {multiInjectedProviders
+          // Even if we detect the MetaMask Extension, we prefere to use the MetaMask SDK
+          .filter((providerInfo) => !providerInfo.info.rdns.startsWith('io.metamask'))
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/wallet/containers/ConnectWalletOptions.tsx:67]: Not 100% sure, but I _think_ this is somewhat related to this:

image

Injected wallets can now get duplicated.
Not reproducible on prod.

Might be related to injected wallets behaving like Metamask.

**Review by alfetopito — COMMENTED** [review]
Hey there, thanks a lot for your contribution!

There are a few issues we have noticed, could you please take a look at them?

One question, I see the SDK is bringing a lot of dependencies.
What's the bundle size it's adding?

**Comment by baptiste-marchand** [comment]
I have read the CLA Document and I hereby sign the CLA

File: libs/wallet/src/web3-react/connection/metaMaskSdk.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+const [web3MetaMask, web3MetaMaskHooks] = initializeConnector<MetaMaskSDK>(
+  (actions) => new MetaMaskSDK({
+    actions, options: {
+      dappMetadata: {
+        name: 'CoW Swap',
+        url: 'https://swap.cow.fi'
+      },
+      readonlyRPCMap: {
+        [`0x${SupportedChainId.MAINNET.toString(16)}`]: RPC_URLS[SupportedChainId.MAINNET]
```
[the remark this answers is not in the document]
Comment (baptiste-marchand) [reply libs/wallet/src/web3-react/connection/metaMaskSdk.tsx]: Fixed, thank you

File: apps/cowswap-frontend/src/modules/wallet/containers/ConnectWalletOptions.tsx (hunk 1 above)
Comment (baptiste-marchand) [reply apps/cowswap-frontend/src/modules/wallet/containers/ConnectWalletOptions.tsx:67]: Fixed. It was related to [this line](https://github.com/cowprotocol/cowswap/pull/5028/files#diff-aebb3e19e3823d5ba6beab9c5724f0ca47417367148900bd3bd0f80e05d971dfR29)

**Comment by baptiste** [comment]
I will not sign the CLA

**Review by alfetopito — APPROVED** [review]
It's working great!

Can I just ask that the original author (@EdouardBougon) sign the CLA too?

**Comment by EdouardBougon** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by anxolin — COMMENTED** [review]
I think the app was already detecting Metmask as per https://eips.ethereum.org/EIPS/eip-6963

My concern with the PR is that its not working great when you use other injected wallet that is not Metamask. 

Also Metamask icon shows first instead of the ones of the available wallets.

Look what we show in production:
image

I am a Rabby wallet user (sorry, I switched some time ago), and this version show the Metamask icon first. We should give more priority to the wallets the user has. If they have Metamask, it will show up.

image

There's a second issue, if I click in "Metamask" it doesn't do anything, it remains connecting forever (not realising I don't have it installed)
image


Another issue has been raised by Leandro above. He is concerned with the bundle size. I think you lazyload, but still would be good to have the dependencies in check, plus having more dependencies adds vulnerabilities (see what happened with Ledger library some months ago). If they are not necessary and we can use EIP standards instead, I would prefer to do so. Ultimately every wallet wants to push for their SDK into the Dapps, but this is not sustainable

**Comment by baptiste-marchand** [comment]
> There's a second issue, if I click in "Metamask" it doesn't do anything, it remains connecting forever (not realising I don't have it installed)
@anxolin I managed to reproduce it only by disabling the MetaMask extension after having clicked on `connect wallet` on Cowswap. Is that also the situation you found yourself in, or it was in a more normal situation?

**Comment by baptiste-marchand** [comment]
About your questions related to the MetaMask SDK, a member from the team should reach out to you on telegram

**Comment by baptiste-marchand** [comment]
> One nitpick related to the uninstalled case: when I close 'connect' modal that appears when MM is not installed, connection modals remains in the 'connecting' state. It would be nice to show a connection error instead (like it works for WC option:

@elena-zh I'm not sure I can fix this one. This is related to how web3-react and MetaMask SDK are integrated together. From what I understood WC is throwing an error when the modal is closed, while MetaMask doesn't.

**Comment by baptiste-marchand** [comment]
Hi @anxolin! I wanted to give you a quick update: our team is actively working on reducing the MetaMask SDK bundle size and removing some dependencies. We will provide you with further updates next week.

**Review by shoom3301 — APPROVED** [review]

**Comment by elena-zh** [comment]
Hey @EdouardBougon , @baptiste-marchand ,
 we were about to include this PR into the release, but, unfortunately, we had to revert these changes in #5125. 
Main problems: 
1. Impossible to sign orders in a mobile device (using WC connector): the app hangs in this state (see the image) and there is no request in MM 
image
2. MM is not installed on a desktop version, I connected to it using a native connector: impossible to sign transactions using it
image
image
3. MM is not installed on a desktop version, I connected to it using regular WC option> each time I pick a token, and try signing an order, this modal pops up

image

**Pull request merged**